File: 100joon/Gold/12865.py
import sys

input = sys.stdin.readline

n, k = map(int, input().split())
dp = [0] * (k + 1)
visited = [set() for _ in range(k + 1)]
bags = []
for i in range(n):
    w, v = map(int, input().split())
    bags.append((i, w, v))
bags.sort(key=lambda x: (x[1], x[2]))


for i in range(1, k + 1):
    for bag in bags:
        idx, w, v = bag
        if i - w < 0:
            break

        q = dp[i - w] + v
        if idx not in visited[i - w] and q > dp[i]:
            dp[i] = q
            visited[i] = visited[i - w] | set([idx])

print(max(dp))

File: 100joon/Gold/12865.js
const input = require("fs")
  .readFileSync("input.txt", "utf-8")
  .trim()
  .split("\n");

const [n, k] = input[0].split(" ").map(Number);
const bags = [];
const visited = Array(k + 1).fill(new Set());
for (let i = 0; i < n; i++) {
  const [w, v] = input[i + 1].split(" ").map(Number);
  bags.push([i, w, v]);
}

bags.sort((a, b) => a[1] - b[1]);

const dp = Array(k + 1).fill(0);
for (let i = 1; i <= k; i++) {
  for (let [idx, w, v] of bags) {
    if (i - w < 0) break;

    const q = dp[i - w] + v;
    if (!visited[i - w].has(idx) && dp[i] < q) {
      visited[i] = new Set([...visited[i - w], idx]);
      dp[i] = q;
    }
  }
}

console.log(Math.max(...dp));
